src/plan_add_hashagg.c

```diff
@@ -0,0 +1,585 @@
+
+	if (!IsA(first, Const))
+		return INVALID_ESTIMATE;
+
+	c = (Const *) first;
+	interval = DatumGetTextPP(c->constvalue);
+	interval_period = (double) date_trunc_interval_period_approx(interval);
+	if (interval_period <= 0)
+		return INVALID_ESTIMATE;
+
+	max_period = estimate_max_spread_expr(root, second);
+	if (!IS_VALID_ESTIMATE(max_period))
+		return INVALID_ESTIMATE;
+
+	return clamp_row_est(max_period / interval_period);
+}
+
+/* if performing integer division number of groups is less than the spread divided by the divisor.
+ * Note that this is an overestimate. */
+static double
+custom_group_estimate_integer_division(PlannerInfo *root, Oid opno, Node *left, Node *right)
+{
+	char	   *function_name = get_opname(opno);
+
+	/* only handle division */
+	if (strlen(function_name) == 1 && function_name[0] == '/' && IsA(right, Const))
+	{
+		Const	   *c = (Const *) right;
+		int64		period;
+
+		if (c->consttype != INT2OID && c->consttype != INT4OID && c->consttype != INT8OID)
+			return INVALID_ESTIMATE;
+
+		period = estimate_max_spread_expr(root, (Expr *) left);
+		if (IS_VALID_ESTIMATE(period))
+			return period / c->constvalue;
+	}
+	return INVALID_ESTIMATE;
+}
+
+/* Get a custom estimate for the number of groups of an expression. Return INVALID_ESTIMATE if we don't have
+ * any extra knowledge and should just use the default estimate */
+static double
+custom_group_estimate_expr(PlannerInfo *root, Node *expr, double path_rows)
+{
+	if (IsA(expr, FuncExpr))
+	{
+		FuncExpr   *custom_group_estimate_func = (FuncExpr *) expr;
+		CustomEstimateForFunctionInfo *hafd = get_custom_estimate_func_info(custom_group_estimate_func->funcid);
+
+		if (NULL != hafd)
+			return hafd->custom_group_estimate_func(root, custom_group_estimate_func, path_rows);
+	}
+	else if (IsA(expr, OpExpr))
+	{
+		OpExpr	   *opexpr = (OpExpr *) expr;
+		Node	   *first;
+		Node	   *second;
+		double		estimate;
+
+		if (list_length(opexpr->args) != 2)
+			return INVALID_ESTIMATE;
+
+		first = eval_const_expressions(root, linitial(opexpr->args));
+		second = eval_const_expressions(root, lsecond(opexpr->args));
+
+		estimate = custom_group_estimate_integer_division(root, opexpr->opno, first, second);
+		if (IS_VALID_ESTIMATE(estimate))
+			return estimate;
+
+		if (IsA(first, Const))
+			return custom_group_estimate_expr(root, second, path_rows);
+		if (IsA(second, Const))
+			return custom_group_estimate_expr(root, first, path_rows);
+	}
+
+	return INVALID_ESTIMATE;
+}
+
+/* Get a custom estimate for the number of groups in a query. Return INVALID_ESTIMATE if we don't have
+ * any extra knowledge and should just use the default estimate. This works by getting
+ * a custom estimate for any groups where a custom estimate exists and multiplying that
+ * by the standard estimate of the groups for which custom estimates don't exist */
+static double
+custom_group_estimate(PlannerInfo *root,
+					  double path_rows)
+{
+	Query	   *parse = root->parse;
+	double		dNumGroups = 1;
```

Comment:
Fix name formatting of variables, here and elsewhere. For instance `dNumGroups` => `d_num_groups`

Reply:
Fixed